Webpack 2 dynamic async imports resolve subdirs incorrectly

**Do you want to request a *feature* or report a *bug*?**
I wish to report a bug.

**What is the current behavior?**

When you run 

```
const dirname = 'MyModule';
System.import(`./subdir/${dirname}`)
```

It resolves the file path to `./MyModule` and says the module cannot be found.

**If the current behavior is a bug, please provide the steps to reproduce.**

Here is a repository that reproduces this behavior. It also shows that the dynamic path resolution does work when subdirs and implicit index.js are not involved.

https://github.com/Swizec/webpack-bug-showcase/blob/dynamic-async-path-bug/src/index.js

**What is the expected behavior?**

Webpack loads `./subdir/MyModule/index.js`

**Please mention other relevant information such as the browser version, Node.js version, Operating System and programming language.**

Webpack 2.2.0.rc.3
node 6.9.1
Chrome 55.0.2883.95 (stable)
Babel 6.12

It's the same with `import()`:
```javascript
const sameModule = await import('./deps/a-module');
sameModule.subModule(); // outputs 'now in sub module'
    
const sameModuleName = 'a-module';
const { subModule } = await import(`./deps/${name}`);  // throws Uncaught (in promise) Error: Cannot find module './'.
subModule();
````

--- a/lib/WebpackOptionsApply.js
+++ b/lib/WebpackOptionsApply.js
@@ -247,7 +247,7 @@ class WebpackOptionsApply extends OptionsApply {
 			new UseStrictPlugin(),
 			new RequireIncludePlugin(),
 			new RequireEnsurePlugin(),
-			new RequireContextPlugin(options.resolve.modules, options.resolve.extensions),
+			new RequireContextPlugin(options.resolve.modules, options.resolve.extensions, options.resolve.mainFiles),
 			new ImportPlugin(options.module),
 			new SystemPlugin(options.module)
 		);

--- a/lib/WebpackOptionsDefaulter.js
+++ b/lib/WebpackOptionsDefaulter.js
@@ -84,6 +84,7 @@ class WebpackOptionsDefaulter extends OptionsDefaulter {
 		this.set("resolve.unsafeCache", true);
 		this.set("resolve.modules", ["node_modules"]);
 		this.set("resolve.extensions", [".js", ".json"]);
+		this.set("resolve.mainFiles", ["index"]);
 		this.set("resolve.aliasFields", "make", (options) => {
 			if(options.target === "web" || options.target === "webworker")
 				return ["browser"];
@@ -100,6 +101,7 @@ class WebpackOptionsDefaulter extends OptionsDefaulter {
 		this.set("resolveLoader.unsafeCache", true);
 		this.set("resolveLoader.mainFields", ["loader", "main"]);
 		this.set("resolveLoader.extensions", [".js", ".json"]);
+		this.set("resolveLoader.mainFiles", ["index"]);
 	}
 }
 

--- a/lib/dependencies/RequireContextPlugin.js
+++ b/lib/dependencies/RequireContextPlugin.js
@@ -10,18 +10,17 @@ const ContextElementDependency = require("./ContextElementDependency");
 const RequireContextDependencyParserPlugin = require("./RequireContextDependencyParserPlugin");
 
 class RequireContextPlugin {
-	constructor(modulesDirectories, extensions) {
+	constructor(modulesDirectories, extensions, mainFiles) {
 		if(!Array.isArray(modulesDirectories))
 			throw new Error("modulesDirectories must be an array");
 		if(!Array.isArray(extensions))
 			throw new Error("extensions must be an array");
 		this.modulesDirectories = modulesDirectories;
 		this.extensions = extensions;
+		this.mainFiles = mainFiles;
 	}
 
 	apply(compiler) {
-		const modulesDirectories = this.modulesDirectories;
-		const extensions = this.extensions;
 		compiler.plugin("compilation", (compilation, params) => {
 			const contextModuleFactory = params.contextModuleFactory;
 			const normalModuleFactory = params.normalModuleFactory;
@@ -43,7 +42,7 @@ class RequireContextPlugin {
 				if(items.length === 0) return callback(null, items);
 
 				callback(null, items.map((obj) => {
-					return extensions.filter((ext) => {
+					return this.extensions.filter((ext) => {
 						const l = obj.request.length;
 						return l > ext.length && obj.request.substr(l - ext.length, l) === ext;
 					}).map((ext) => {
@@ -60,8 +59,28 @@ class RequireContextPlugin {
 				if(items.length === 0) return callback(null, items);
 
 				callback(null, items.map((obj) => {
-					for(let i = 0; i < modulesDirectories.length; i++) {
-						const dir = modulesDirectories[i];
+					return this.mainFiles.filter((mainFile) => {
+						const l = obj.request.length;
+						return l > mainFile.length + 1 && obj.request.substr(l - mainFile.length - 1, l) === "/" + mainFile;
+					}).map((mainFile) => {
+						const l = obj.request.length;
+						return [{
+							context: obj.context,
+							request: obj.request.substr(0, l - mainFile.length)
+						}, {
+							context: obj.context,
+							request: obj.request.substr(0, l - mainFile.length - 1)
+						}];
+					}).reduce((a, b) => a.concat(b), []).concat(obj);
+				}).reduce((a, b) => a.concat(b), []));
+			});
+
+			params.contextModuleFactory.plugin("alternatives", (items, callback) => {
+				if(items.length === 0) return callback(null, items);
+
+				callback(null, items.map((obj) => {
+					for(let i = 0; i < this.modulesDirectories.length; i++) {
+						const dir = this.modulesDirectories[i];
 						const idx = obj.request.indexOf("./" + dir + "/");
 						if(idx === 0) {
 							obj.request = obj.request.slice(dir.length + 3);

--- a/test/cases/context/issue-3873/index.js
+++ b/test/cases/context/issue-3873/index.js
@@ -1,0 +1,7 @@
+function get(name) {
+	return require("./" + name);
+}
+
+it("should automatically infer the index.js file", function() {
+	get("module").should.be.eql("module");
+});

--- a/test/cases/context/issue-3873/module/index.js
+++ b/test/cases/context/issue-3873/module/index.js
@@ -1,0 +1,1 @@
+module.exports = "module";

